thorlcr/activities/trace/thtraceslave.cpp

```diff
@@ -0,0 +1,159 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "thtraceslave.ipp"
+
+class CTraceSlaveActivity : public CSlaveActivity, public CThorDataLink, public CThorSteppable
+{
+    IThorDataLink *input;
+    IHThorTraceArg *helper;
+    IOutputMetaData *meta;
+    IRangeCompare *stepCompare;
+    OwnedRoxieString name;
+    unsigned keepLimit;
+    unsigned skip;
+    unsigned sample;
+    bool traceEnabled;
+
+public:
+    IMPLEMENT_IINTERFACE_USING(CSimpleInterface);
+
+    CTraceSlaveActivity(CGraphElementBase *_container)
+        : CSlaveActivity(_container), CThorDataLink(this), CThorSteppable(this),
+          meta(NULL), keepLimit(0), skip(0), sample(0), traceEnabled(false)
+    {
+        helper = (IHThorTraceArg *) queryHelper();
+        stepCompare = NULL;
+    }
+    void init(MemoryBuffer &data, MemoryBuffer &slaveData)
+    {
+        helper = (IHThorTraceArg *)queryHelper();
+        appendOutputLinked(this);
+    }
+    void start()
+    {
+        ActivityTimer s(totalCycles, timeActivities);
+        dataLinkStart();
+        input = inputs.item(0);
+        startInput(input);
+        IInputSteppingMeta *stepMeta = input->querySteppingMeta();
+        if (stepMeta)
+            stepCompare = stepMeta->queryCompare();
+        meta = helper->queryOutputMeta();
+        traceEnabled = getOptBool(THOROPT_TRACE_ENABLED, false);
+        keepLimit = helper->getKeepLimit();
+        if (keepLimit==(unsigned) -1)
+            keepLimit = getOptUInt(THOROPT_TRACE_LIMIT, 100);
+        if (!traceEnabled)
+            keepLimit = 0;
+        if (!helper->canMatchAny())
+            traceEnabled = false;
+        skip = helper->getSkip();
+        sample = helper->getSample();
+        if (sample)
+            sample--;
```

Comment:
If sample is e.g. 1, it's supposed to mean every row, or every other row?
As it stands, the 1st row will be skipped by the initial sample check:
```
            else if (sample)
                sample--;
```
but when reset, will decrement immediately here and output all others...